### ISSUE TITLE
copy drops setuid bit

### ISSUE THREAD

**Issue by joejulian** [projectatomic/buildah#574]
**Description**

When I issue a buildah copy of a directory tree created using debootstrap, the setuid bit is dropped causing sudo to not work.

I can work around this with a `buildah mount`, `cp -a`, `buildah umount`

**Steps to reproduce the issue:**
1. create a directory tree with files that have the setuid bit set
2. buildah copy that tree

**Describe the results you received:**
setuid bits were lost

**Describe the results you expected:**
setuid bits would be retained

**Output of `rpm -q buildah` or `apt list buildah`:**

This is not installed by package manager (no package for Arch).

**Output of `buildah version`:**

```
Version:       0.15
Go Version:    go1.10
Image Spec:    1.0.0
Runtime Spec:  1.0.0
Git Commit:    cd999fa
Built:         Tue Mar 27 10:28:54 2018
OS/Arch:       linux/amd64
```

**Output of `cat /etc/*release`:**

```
LSB_VERSION=1.4
DISTRIB_ID=Arch
DISTRIB_RELEASE=rolling
DISTRIB_DESCRIPTION="Arch Linux"
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
ID_LIKE=archlinux
ANSI_COLOR="0;36"
HOME_URL="https://www.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
```

**Output of `uname -a`:**

```
Linux questor 4.14.13-1-ARCH #1 SMP PREEMPT Wed Jan 10 11:14:50 UTC 2018 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
No such file or directory
```

**Comment by TomSweeneyRedHat** [comment]
@joejulian thanks for the problem report, we'll take a look at it.

**Comment by rhatdan** [comment]
@joejulian Do you know if this works correctly in a Dockerfile? Using `docker build`

**Comment by rhatdan** [comment]
@nalind Any ideas?

**Comment by pixdrift** [comment]
I have reproduced to test if it was an issue with `var` being mounted nosuid, but issue persists when nosuid isn't configured.

The debug output is the following:
`DEBU[0000] CopyFileWithTar(/bin/ping, /var/lib/containers/storage/overlay/53039fe01d8fee33b79b9ff83d95a74e49e36d07b83decf884c5086be0906339/merged/bin/ping)`

I suspect the issue may be here when the file mode is copied:
https://github.com/containers/storage/blob/5d52f079f1709f0408a6c086c603d259b0e5da3e/pkg/archive/archive.go#L1137-L1142

**Comment by pixdrift** [comment]
@nalind can you take a look at this?
https://github.com/containers/storage/blame/5d52f079f1709f0408a6c086c603d259b0e5da3e/pkg/archive/archive.go#L1142

It looks like this line is a fix for Windows.. which may be modifying the `hdr.Mode` set by `FileInfoHeader` just above? Thoughts?

**Comment by rhatdan** [comment]
Well on Linux
```
func chmodTarEntry(perm os.FileMode) os.FileMode {
	return perm // noop for unix as golang APIs provide perm bits correctly
}
```
chmodTarEntry is a noop.

**Comment by rhatdan** [comment]
Which leaves us with
os.FileMode(hdr.Mode)

Looks like a bug in go.

```
cat /tmp/t.go
package main

import (
	"fmt"
	"os"
)

func main() {
	mode := int(0755)
	fmt.Println(os.FileMode(mode))
	mode = int(04755)
	fmt.Println(os.FileMode(mode))
}
```

go run /tmp/t.go 
-rwxr-xr-x
-rwxr-xr-x

**Comment by pixdrift** [comment]
It looks like it may be here in the go source
https://github.com/golang/go/blob/master/src/os/types.go#L66-L74

Can't get it to pick up setuid here

**Comment by pixdrift** [comment]
It looks like `hdr.Mode` makes it through the process even though `os.FileMode` doesn't display as expected.
```
DEBU[0000] hdr.Mode before: 35309 / 1000100111101101 / -rwxr-xr-x 
DEBU[0000] hdr.Mode after: 35309 / 1000100111101101 / -rwxr-xr-x 
```

**Comment by rhatdan** [comment]
Well that would mean the untar should work, or we were looking at a red herring?

**Comment by joejulian** [comment]
@rhatdan No,. I did not happen to know but just tested. Docker's copy also does not retain setuid.

**Comment by rhatdan** [comment]
@pixdrift Could you see if you can use that mode to change the mode on a file to setuid?

**Comment by pixdrift** [comment]
@rhatdan not 100% clear on your suggestion, but the binary going into the hdr.Mode suggests that the result should be setuid. I tried to follow it through the untar process but had trouble determining where the execution went. I notice that the untar process takes options and was wondering if the untar process itself was stripping the setuid bit.

I tested copying setuid file into the mounted container using `cp -a` (@joejulian also mentioned this above), the setuid bit survives, so it's definitely something in the buildah cp.

I must admit I am stuck determining where it is dropped in the archive code, and if dropping it is intentional. As @joejulian has mentioned, this looks like it is standard docker behaviour too, but I am intrigued as to why.

hdr contents looks ok when it hits here, it's then written out by tar.NewWriter
https://github.com/containers/storage/blob/5d52f079f1709f0408a6c086c603d259b0e5da3e/pkg/archive/archive.go#L1148

Not sure what the etiquette is to ask for someone to assist.. nalind perhaps? (because I see they have contributed elsewhere in the archiver code)

**Comment by rhatdan** [comment]
@vbatts @runcom WDYT?

**Comment by vbatts** [comment]
hmm. I'm not sure that it's in the archive layers, but might be. I think the `os.FileMode` behavior smells funny. 

```golang
package main

import (
        "fmt"
        "os"
)

func main() {
        mode := int(0755)
        fmt.Println(os.FileMode(mode))
        fmt.Println((mode))
        mode = int(04755)
        fmt.Println(os.FileMode(mode))
        fmt.Println((mode))
}
```

```
$ go run f.go
-rwxr-xr-x
493
-rwxr-xr-x
2541
```

**Comment by rhatdan** [comment]
We should open an issue with Golang for this.

**Comment by optechdvb** [comment]
This issue is still present on 18.06.1-ce 

container1:latest has embedded target rootfs image.

FROM container1:latest as sdk

FROM ubuntu
COPY from=sdk /target-roots/ /target-rootfs/

SUID and SGID bits not observed in resulting container


Has this been pursued further with Golang? Has a defect been filed?

**Comment by rhatdan** [comment]
No update on this one.

**Comment by rhatdan** [comment]
Any chance newer versions of golang fixed this?

**Comment by rhatdan** [comment]
go run /tmp/t.go 
-rwxr-xr-x
493
-rwxr-xr-x
2541

I guess not

**Comment by rhatdan** [comment]
go version
go version go1.12.5 linux/amd64

**Comment by renich** [comment]
I am suffering from this as well. Can't use sudo due to this.

**Comment by jamescassell** [comment]
I also ran into this one... without this fixed, making minimal images with multi-stage Dockerfile's is broken due to incorrect file capabilities. (Didn't test docker, but sounds like it's also broken.)

**Comment by rhatdan** [comment]
@nalind Is there a way we could special case this?

**Comment by nalind** [comment]
My experience with `docker build` CE 19.03.12 is that the setuid and setgid bits are stripped from content copied from the build context (the sticky bit, which is nearby, is preserved).  The setuid and setgid bits are preserved when content is copied from another build stage with "COPY --from", or extracted from an archive by the "ADD" instruction.

I think 18.06 from Fedora's moby-engine package preserved them in all cases.  We're running conformance tests using 19.03, though, and it's the newest CE version I can find, so we're probably going to have to match the 19.03 behavior for builds that use Dockerfiles, even though I personally think that always preserving them is the more correct thing to do.

But for `buildah add`, we can default to preserving them, since it doesn't have a notion of a build context directory.  Self-assigning, since I'm already in that area.
